packages/dev/core/src/Materials/Node/Blocks/Dual/TBNBlock.ts

```diff
@@ -0,0 +1,171 @@
+        );
+    }
+
+    /**
+     * Gets the current class name
+     * @returns the class name
+     */
+    public getClassName() {
+        return "TBNBlock";
+    }
+
+    /**
+     * Initialize the block and prepare the context for build
+     * @param state defines the state that will be used for the build
+     */
+    public initialize(state: NodeMaterialBuildState) {
+        state._excludeVariableName("tbnNormal");
+        state._excludeVariableName("tbnTangent");
+        state._excludeVariableName("tbnBitangent");
+    }
+
+    /**
+     * Gets the normal input component
+     */
+    public get normal(): NodeMaterialConnectionPoint {
+        return this._inputs[0];
+    }
+
+    /**
+     * Gets the tangent input component
+     */
+    public get tangent(): NodeMaterialConnectionPoint {
+        return this._inputs[1];
+    }
+
+    /**
+     * Gets the world matrix input component
+     */
+    public get world(): NodeMaterialConnectionPoint {
+        return this._inputs[2];
+    }
+
+    /**
+     * Gets the TBN output component
+     */
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    public get TBN(): NodeMaterialConnectionPoint {
+        return this._outputs[0];
+    }
+
+    public get target() {
+        return NodeMaterialBlockTargets.VertexAndFragment;
+    }
+
+    public set target(value: NodeMaterialBlockTargets) {}
+
+    public autoConfigure(material: NodeMaterial) {
+        if (!this.world.isConnected) {
+            let worldInput = material.getInputBlockByPredicate((b) => b.isSystemValue && b.systemValue === NodeMaterialSystemValues.World);
+
+            if (!worldInput) {
+                worldInput = new InputBlock("world");
+                worldInput.setAsSystemValue(NodeMaterialSystemValues.World);
+            }
+            worldInput.output.connectTo(this.world);
+        }
+
+        if (!this.normal.isConnected) {
+            let normalInput = material.getInputBlockByPredicate((b) => b.isAttribute && b.name === "normal");
+
+            if (!normalInput) {
+                normalInput = new InputBlock("normal");
+                normalInput.setAsAttribute("normal");
+            }
+            normalInput.output.connectTo(this.normal);
+        }
+
+        if (!this.tangent.isConnected) {
+            let tangentInput = material.getInputBlockByPredicate((b) => b.isAttribute && b.name === "tangent" && b.type === NodeMaterialBlockConnectionPointTypes.Vector4);
+
+            if (!tangentInput) {
+                tangentInput = new InputBlock("tangent");
+                tangentInput.setAsAttribute("tangent");
+            }
+            tangentInput.output.connectTo(this.tangent);
+        }
+    }
+
+    public prepareDefines(mesh: AbstractMesh, nodeMaterial: NodeMaterial, defines: NodeMaterialDefines) {
+        const normal = this.normal;
+        const tangent = this.tangent;
+
+        let normalAvailable = normal.isConnected && normal.isConnectedToInputBlock;

```

Comment:
Why testing that the normal is coming only from an `InputBlock`? Shouldn't we simply use `let normalAvailable = normal.isConnected;`?

Reply:
Good catch